plugins/ata/fu-ata-device.c

```diff
@@ -0,0 +1,603 @@
+	g_string_append_printf (str, "    transfer-mode:\t0x%x\n", (guint) self->transfer_mode);
+	g_string_append_printf (str, "    transfer-size:\t0x%x\n", (guint) self->transfer_size);
+	g_string_append_printf (str, "    pci-depth:\t\t%u\n", self->pci_depth);
+}
+
+/* https://docs.microsoft.com/en-us/windows-hardware/drivers/install/identifiers-for-ide-devices */
+static gchar *
+fu_ata_device_pad_string_for_id (const gchar *name)
+{
+	GString *str = g_string_new (name);
+	fu_common_string_replace (str, " ", "_");
+	for (guint i = str->len; i < 40; i++)
+		g_string_append_c (str, '_');
+	return g_string_free (str, FALSE);
+}
+
+static gboolean
+fu_ata_device_parse_id (FuAtaDevice *self, const guint8 *buf, gsize sz, GError **error)
+{
+	FuDevice *device = FU_DEVICE (self);
+	guint64 xfer_min = 1;
+	guint64 xfer_max = 0xffff;
+	guint16 id[FU_ATA_IDENTIFY_SIZE/2];
+	g_autofree gchar *name_pad = NULL;
+	g_autofree gchar *sku = NULL;
+
+	/* check size */
+	if (sz != FU_ATA_IDENTIFY_SIZE) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "ID incorrect size, got 0x%02x",
+			     (guint) sz);
+		return FALSE;
+	}
+
+	/* read LE buffer */
+	for (guint i = 0; i < sz / 2; i++)
+		id[i] = fu_common_read_uint16 (buf + (i * 2), G_LITTLE_ENDIAN);
+
+	/* verify drive correctly supports DOWNLOAD_MICROCODE */
+	if (!(id[83] & 1 && id[86] & 1)) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "DOWNLOAD_MICROCODE not supported by device");
+		return FALSE;
+	}
+
+	/* fallback to a sane default */
+	if (self->transfer_mode == 0) {
+		if ((id[119] & 0x10) && (id[120] & 0x10))
+			self->transfer_mode = 0x3;
+		else
+			self->transfer_mode = 0x7;
+	}
+
+	/* the newer, segmented transfer mode */
+	if (self->transfer_mode == 0x3 || self->transfer_mode == 0xe) {
+		xfer_min = id[234];
+		if (xfer_min == 0x0 || xfer_min == 0xffff)
+			xfer_min = 1;
+		xfer_max = id[235];
+		if (xfer_max == 0x0 || xfer_max == 0xffff)
+			xfer_max = xfer_min;
+	}
+
+	/* fall back to a sane block size */
+	if (self->transfer_size == 0x0)
+		self->transfer_size = xfer_min * FU_ATA_BLOCK_SIZE;
+	else if (self->transfer_size == 0xffff)
+		self->transfer_size = xfer_max * FU_ATA_BLOCK_SIZE;
+
+	/* get values in case the kernel didn't */
+	if (fu_device_get_serial (device) == NULL) {
+		g_autofree gchar *tmp = NULL;
+		tmp = fu_ata_device_get_string (id, 10, 19);
+		fu_device_set_serial (device, tmp);
+	}
+	if (fu_device_get_name (device) == NULL) {
+		g_autofree gchar *tmp = NULL;
+		tmp = fu_ata_device_get_string (id, 27, 46);
+		fu_device_set_name (device, tmp);
+	}
+	if (fu_device_get_version (device) == NULL) {
+		g_autofree gchar *tmp = NULL;
+		tmp = fu_ata_device_get_string (id, 23, 26);
+		fu_device_set_version (device, tmp);
+	}
+
+	/* 8 byte additional product identifier == SKU? */
+	sku = fu_ata_device_get_string (id, 170, 173);
+	if (sku != NULL)
+		g_debug ("SKU=%s", sku);
```

Comment:
Hopefully someone at a storage vendor can validate this.

Reply:
Valid point. Fixed.

Reply:
It's in the spec, but I've not found any hardware that supports it. I've asked the Phison for more details -- even if it's set I'm not quite sure how to use it for GUID generation. It's more of a curiosity for now.

Reply:
Well the same thing Dell does for GUID magic with the identify block in NVME also exists in ATA.  I don't have a disk with it though, so I'll have to inquire about adding it later after this plugin is merged.

Reply:
Great, thanks. When you do, it'll need the efivar dep added too.